x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/DataFrameDataExtractorFactory.java

```diff
@@ -182,10 +189,37 @@ private static void validateIndexAndExtractFields(Client client,
             listener::onFailure
         );
 
-        // Step 2. Get field capabilities necessary to build the information of how to extract fields
-        ActionListener<Integer> docValueFieldsLimitListener = ActionListener.wrap(
-            docValueFieldsLimit -> {
-                docValueFieldsLimitHolder.set(docValueFieldsLimit);
+        // Step 3. Get field capabilities necessary to build the information of how to extract fields
+        ActionListener<SearchResponse> checkCardinalityHandler = ActionListener.wrap(
+            searchResponse -> {
+                Map<String, Long> fieldCardinalityLimits = config.getAnalysis().getFieldCardinalityLimits();
+                if (fieldCardinalityLimits.isEmpty() == false) {
+                    if (searchResponse == null) {
+                        listener.onFailure(ExceptionsHelper.badRequestException("searchResponse == null"));
+                        return;
+                    }
+                    Aggregations aggs = searchResponse.getAggregations();
+                    if (aggs == null) {
+                        listener.onFailure(ExceptionsHelper.badRequestException("aggs == null"));
+                        return;
+                    }
+                    for (Map.Entry<String, Long> entry : fieldCardinalityLimits.entrySet()) {
+                        String fieldName = entry.getKey();
+                        Long limit = entry.getValue();
+                        Cardinality cardinality = aggs.get(fieldName);
+                        if (cardinality == null) {
+                            listener.onFailure(ExceptionsHelper.badRequestException("cardinality == null"));
+                            return;
+                        }
+                        if (cardinality.getValue() > limit) {
+                            listener.onFailure(
+                                ExceptionsHelper.badRequestException(
+                                    "Field [{}] must have at most [{}] distinct values but there were [{}]",
```

Comment:
```suggestion
                                    "Field [{}] must have at most [{}] distinct values but there were at least [{}]",
```
Cardinality is approximate. So, we should give the number it returned as the bottom limit.

Reply:
Done.